File: cc/044.cc
// Project Euler Problem 044
//
// Find the smallest pair of pentagonal numbers whose sum and
// difference is pentagonal.

#include <cstdio>
#include <vector>
#include <set>

using namespace std;

void generate_to(long limit, size_t length, vector<long>& v, set<long>& s) {
  size_t n = v.size() + 1;
  long p;
  do {
    p = n * (3*n-1) / 2;
    v.push_back(p);
    s.insert(p);
    n++;
  } while (p < limit || n <= length);
}

int main(int argc, const char* argv[]) {
  vector<long> v;
  set<long> s;
  generate_to(100, 10, v, s);
  long n2 = 1;
  while (true) {
    n2++;
    generate_to(1, n2, v, s);
    long p2 = v[n2];
    for (long n1=0; n1<n2; n1++) {
      long p1 = v[n1];
      long diff = p2-p1;
      if (!s.count(diff)) continue;
      long sum = p2+p1;
      generate_to(sum, 1, v, s);
      if (s.count(sum)) {
        printf("%ld\n", diff);
        exit(0);
      }
    }
  }
}

File: go/problems/044.go
// Project Euler Problem 044
//
// Find the smallest pair of pentagonal numbers whose sum and
// difference is pentagonal.

package problems

import (
	"fmt"
	"math"
)

func isPentagonal(p int) bool {
	n := int(math.Sqrt(float64(2 * p / 3)))
	for ; ; n++ {
		pp := n * (3*n - 1) / 2
		if pp > p {
			return false
		}
		if pp == p {
			return true
		}
	}
	panic("impossible!")
}

func Problem044() string {
	for k := 1; ; k++ {
		pk := k * (3*k - 1) / 2
		for j := 1; j < k; j++ {
			pj := j * (3*j - 1) / 2
			if isPentagonal(pk-pj) && isPentagonal(pk+pj) {
				return fmt.Sprintf("%d", pk-pj)
			}
		}
	}
	panic("Cannot reach")
}

func init() {
	Functions[44] = Problem044
}
